ref: Avoid the encoding crate by using WStr and Utf16Error

flub opened the pull request:
This re-models the Utf16Error after std::str::Utf8Error which allows
it to provide enough information to write a decoder.  This means we no
longer need a decoder and can write our segmentation iterator a little
simpler.

The complexity of this is a lot less than using the RawDecoder API
plus there is fewer duplication in the branches, so this is more
maintainable.  Downside is that this has had less production scrutiny.


----

This also fixes a minor bug where we counted the bytes length instead
of character count for deciding if the string is long enough.

#skip-changelog

Commit:
ref: Avoid the encoding crate by using WStr and Utf16Error

This re-models the Utf16Error after std::str::Utf8Error which allows
it to provide enough information to write a decoder.  This means we no
longer need a decoder and can write our segmentation iterator a little
simpler.

The complexity of this is a lot less than using the RawDecoder API
plus there is fewer duplication in the branches, so this is more
maintainable.  Downside is that this has had less production scrutiny.

jan-auer commented on relay-general/src/pii/attachments.rs:
```diff
@@ -269,75 +267,58 @@ impl StringMods for [u8] {
-                    return Some(WstrSegment { encoded, decoded });
-                } else {
-                    continue;
+            let slice: &mut [u8] = match WStr::from_utf16le_mut(&mut self.data[self.offset..]) {
```
The compiler should be able to infer this.

```suggestion
            let slice = match WStr::from_utf16le_mut(&mut self.data[self.offset..]) {
```

flub replied:
uhm, yes.  this was helpful while i was getting the returned types right but forgot to clean up

jan-auer commented on relay-wstring/src/lib.rs:
```diff
@@ -42,15 +45,17 @@ impl fmt::Display for Utf16Error {
+    ///
+    /// If `Some` you should be able to attempt to parse again from the offset given by
+    /// adding the error legnth to [Self::valid_up_to].
+    pub fn error_len(&self) -> Option<usize> {
```
Can you copy over semantics and description from [`Utf8Error::error_len`](https://doc.rust-lang.org/stable/std/str/struct.Utf8Error.html#method.error_len)? When reading this description, I did not understand why there could be `None` and why `Some`.

jan-auer commented on relay-wstring/src/lib.rs:
```diff
@@ -42,15 +45,17 @@ impl fmt::Display for Utf16Error {
+    /// Return the length of the error if it might be recoverable.
+    ///
+    /// If `Some` you should be able to attempt to parse again from the offset given by
+    /// adding the error legnth to [Self::valid_up_to].
```
```suggestion
    /// adding the error length to [Self::valid_up_to].
```

jan-auer commented on relay-wstring/src/lib.rs:
```diff
@@ -434,21 +439,58 @@ unsafe fn decode_surrogates(u: u16, u2: u16) -> char {
+                    if !is_trailing_surrogate(u2) {
+                        return Err(Utf16Error {
+                            valid_up_to: code_point_ptr - base_ptr,
+                            error_len: chunks.next().map(|_chunk| std::mem::size_of::<u16>() as u8),
```
The error len can statically be `2` here, since `u2` is a valid UTF-16 starting character. `None` is invalid here.

Imagine this sequence: `[single] - [single] - [leading] - [single]`. Our error should have `valid_up_to = 4` and `error_len = 2`, because you can parse a valid UTF-16 string for `6..8`.

jan-auer replied:
On rethinking this, you could also be more elaborate (although I'm not sure if that makes sense) and then look even further into the future, but you will have to differentiate based on what the next character is. 

However, the intent behind `error_len` is to print a box character. And in that sense, you may want to print another box character after this, if you, for instance, get another leading character immediately afterwards.

flub replied:
hmm, there is no guarantee that `error_len` is only allowed to cover only one char is there?  Interestingly there *is* a maximum number of bytes in case of `None` according to the description.  So maybe there is this assumption and we can't optimise looking forward further.  The fact that `Utf8Error` stores their `error_len` in a `u8` also re-enforces this assumption i guess.

jan-auer replied:
> there is no guarantee that error_len is only allowed to cover only one char is there?

Looks like there is no such guarantee. However, when reading the description, it sounds like this should cover one replacement character. I'm assuming, for UTF-8 `error_len` will always be in `1..=3`.

Following up with my initial comment: The only edge case why `chunks.next().map()` is asking too much is because it would swallow the last code point in the above example. Imagine the above string renders as `AB�D` and the replacement character is due to a leading surrogate without its trailing counterpart. Your code would probably render as `AB` or `AB�`, since your Error has `valid_up_to=4` and `error_len=None` (due to chunks.next() being None).

flub replied:
Agreed the intention is to be able to insert a replacement char for each error, so we can't skip ahead and combine errors.

So your original comment is right, a lone trailing surrogate can be hardcoded to `error_len=2`.  Also so for a lone leading surrogate.  So we can never return an `error_len` different that 2.  Only a lone leading surrogate at the end of the slice can return `None`.

I don't follow your last example, I think this would result in `AB�D`.

jan-auer commented on relay-general/src/pii/attachments.rs:
```diff
@@ -269,75 +267,58 @@ impl StringMods for [u8] {
+            if slice.len() < MIN_STRING_LEN * std::mem::size_of::<u16>() {
+                // Quick shortcut, actual char count check done below
+                continue;
+            }
```
Is this check really worth it? Lines 311-315 are so cheap that I wouldn't even bother performing this check.

flub replied:
yeah, i wondered this too.  more importantly the `.count()` below is too expensive, fixed that as well.

jan-auer commented on relay-general/src/pii/attachments.rs:
```diff
@@ -269,75 +267,58 @@ impl StringMods for [u8] {
+                Err(err) => {
+                    let start = self.offset;
+                    match err.error_len() {
+                        Some(len) => self.offset += err.valid_up_to() + len,
```
A nice optimization here would be to scan forward (peek?) while the character is a trailing surrogate. That would require knowing about surrogates, however, so I think the current implementation is fine, too.

flub replied:
That optimisation might be possible in the Utf16Error though.

jan-auer replied:
You're right. In the first check for `is_trailing_surrogate` when you don't expect one, you can keep going while you keep receiving another.

flub replied:
So, the other discussion resulted in the conclusion that the API there does not allow this optimisation.  We either re-implement that logic here or we don't do it.  My vote would be to not bother with this unless we have profiling information.

Commit:
Update relay-wstring/src/lib.rs

Co-authored-by: Jan Michael Auer <mail@jauer.org>

Commit:
Remove type annotation

Co-authored-by: Jan Michael Auer <mail@jauer.org>

Commit:
Optimise length checking a bit more

The shortcut probably doesn't do much.

But much more important is that we don't need to consume the entire
iterator to just count to ~5.  Only iterate through something short.

Commit:
Better docs

jan-auer commented on relay-general/src/pii/attachments.rs:
```diff
@@ -314,7 +310,7 @@ impl<'a> Iterator for WStrSegmentIter<'a> {
             };
 
-            if encoded.chars().count() < MIN_STRING_LEN {
+            if encoded.chars().take(MIN_STRING_LEN).count() < MIN_STRING_LEN {
```
Note that this is potentially slower than just doing `.count()` since you have an optimized count implementation that does not decode.

flub replied:
ugh.  can we have a `.count_up_to(5)` iterator method? ;)

but this is a hard call, in general my guess is that fully decoding 5 characters is faster that partially decoding 100+ chars and that we're more likely to have 100+ chars here.

jan-auer replied:
You can keep this implementation. In most cases where long strings are detected, doing `.take().count()` will be faster due to the early exit. I am a bit curious if the compiler can see through the inlined `.count()`, but it's probably not worth going that deep. 

Sorry for the distraction :)

Swatinem commented on relay-general/src/pii/attachments.rs:
```diff
@@ -613,7 +589,13 @@ mod tests {
     fn utf16le(s: &str) -> Vec<u8> {
-        UTF_16LE.encode(s, EncoderTrap::Strict).unwrap()
+        s.encode_utf16()
+            .map(|u| u.to_le_bytes())
```
`.flat_map(|u| u.to_le_bytes().into_iter())` or something along that line should avoid the double collect here…

flub replied:
I tried this again and I'm not sure it's possible.  Problem is that `u.to_le_bytes()` creates a new array, which does not live anywhere, so you can't just turn it into an iterator.  By doing the `collect::<Vec<_>>()` you give it somewhere it lives long enough for the actual flatten.

Swatinem commented on relay-wstring/src/lib.rs:
```diff
@@ -42,15 +45,24 @@ impl fmt::Display for Utf16Error {
+    ///
+    /// If `None`: the end of the input was reached unexpectedly.  `self.valid_up_to()` is 1
+    /// to 3 bytes from the end of the input.  If a byte stream such as a file or a network
+    /// socket is being decoded incrementall, this could still be a valid char whose byte
```
```suggestion
    /// socket is being decoded incrementally, this could still be a valid char whose byte
```

Commit:
Fix typo

Co-authored-by: Arpad Borsos <arpad.borsos@sentry.io>

Commit:
Optimise error Utf16Error returns, clarify doc comments

Commit:
Merge branch 'master' into ref/avoid-encoding-crate

Commit:
Minor readability improvement

Commit:
Someone decided that negation is easier to read than comparison

I'm not sure I'm entirely in agreement with clippy here.  But clippy
is the boss.
